servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/MpmcSequentialRunQueue.java

```diff
@@ -0,0 +1,185 @@
+
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+/**
+ * This queue allows for Muli-Producer Multi-Consumer (Mpmc) threading semantics while also invoking the head
+ * {@link Node}'s {@link Node#run()} in a serial fashion. The {@link Node#run()} will eventually trigger a
+ * {@link #poll(Node)} (possibly asynchronously on another thread) which will invoke {@link Node#run()} on the next head
+ * (assuming one exists).
+ * <p>
+ * Below is the expected interaction pattern and lifecycle of a Node:
+ *     <pre>{@link #offer(Node)} -> {@link Node#run()} -> {@link #poll(Node)}</pre>
+ * <p>
+ * Although this queue supports Multi-Consumer threading semantics the {@link #poll(Node)} is typically only invoked
+ * from a single thread (assuming successful runnable completion), it maybe invoked multiple times (potentially from
+ * different threads) with the same {@link Node} due to cancellation/failure.
+ */
+final class MpmcSequentialRunQueue {
+    private static final AtomicReferenceFieldUpdater<MpmcSequentialRunQueue, Node>
+            tailUpdater = newUpdater(MpmcSequentialRunQueue.class, Node.class, "tail");
+
+    @Nullable
+    private volatile Node tail;
+
+    /**
+     * Offer {@link Node} to this queue.
+     * @param node The {@link Node} to append. Must only exclusively be offered to this queue instance.
+     */
+    void offer(final Node node) {
+        for (;;) {
+            final Node currentTail = this.tail;
+            if (currentTail == null) {
+                if (tailUpdater.compareAndSet(this, null, node)) {
+                    // node has been inserted and is the only node, we initiate processing.
+                    node.run();
+                    break;
+                }
+                // Another thread won the race to offer a node, loop around and try again.
+            } else if (currentTail.append(node)) {
+                // Make the newly appended node visible as the tail. This is a best effort CAS and may fail because:
+                // 1. Another thread is also inserting, has a stale tail, followed its existing tail links, and updated
+                // the tail reference via offerPatchTail.
+                // 2. The consumer thread has seen the link from the old tail to the new node, processed node,
+                // popped node from the list (updated node's next to point to EMPTY_NODE), another producer thread
+                // appends a new node, sees the tail is popped, and updates the tail reference via CAS.
+                tailUpdater.compareAndSet(this, currentTail, node);
+                break;
+            } else if (currentTail.isPopped()) {
+                // A previously appended node was processed, and popped before updating the tail after append. In that
+                // case the tail maybe pointing to an invalid node and we clear it out.
+                if (tailUpdater.compareAndSet(this, currentTail, node)) {
+                    node.run();
```

Comment:
Looks like this can race with `poll()` and run concurrent tasks.


```java
tail = node 1; (node1 currently running)
Thread 1:  
offer (node 2)
 - node1.append(node2) <- success
    - suspended
Thread 2 (concurrent with Thread 1):
  offer(node3)
  fails node1.append() as raced with Thread 1
  suspended

Thread 3:
  node 1 returns; node1.pop() returns node 2;
     node 2.run();
     returns;

Thread 2 resumes:
    node1.isPopped() <- true
    tailUpdater.cas(node1, node3) <- true
      node3.run(); <-- we are running 2 nodes concurrently (Thread 2 => node 3, Thread 3 => node 2)
```

Reply:
great find 🔥🔐! let me just use the simplified queue approach as you suggested offline. that way we can get this in and optimize the queue elements later if necessary.